gnucash/import-export/test/gmock-gnc-prefs.h

```diff
@@ -0,0 +1,47 @@
+#ifndef GMOCK_GNC_PREFS_H
+#define GMOCK_GNC_PREFS_H
+
+#include <gmock/gmock.h>
+
+extern "C"
+{
+#include <gnc-prefs.h>
+#include <gnc-prefs-p.h>
+}
+
+
+// mock up for PrefsBackend (singleton class)
+class MockPrefsBackend : PrefsBackend
+{
+public:
+    MockPrefsBackend(MockPrefsBackend const&) = delete;
+    MockPrefsBackend& operator=(MockPrefsBackend const&) = delete;
```

Comment:
The underlying implementation of the preferences backend follows a singleton pattern by using global pointer `prefsbackend`. Therefore the mock up has to use it as well.

Reply:
Uh, that's not a good reason: The whole point of a mock is to avoid implementation details, and the global variable is just that: An implementation detail.

A better answer would be something along the lines of "gnc_prefs_foo needs some way to access the preferences and since it doesn't have a preferences parameter it needs something internal and a singleton is an easy way to do that." Except that you're not using the singleton for that, you're using a global variable for it and doing so in a poorly thought out way.

The global belongs to gmock-gnc-prefs.c so it could be static with a setter . `MockPrefsBackend` can be an ordinary class with a default constructor and destructor. Let there be `static MockPrefsBackend* prefsbe` and `gnc_prefs_set_backend(MockPrefsBackend*)`. Your fixture code would do
```
   SetUp ()
   {
          gnc_prefs_set_backend(&m_prefs);
          // ...
    }
   MockPrefsBackend m_prefs;
```

Reply:
Oh yes, I haven't thought again about this implementation. I used the singleton pattern in my first approach, where it was really necessary, since I linked *libgnc-core-utils.so* into the test application, which contains *gnc-prefs.c* and defined global pointer `prefsbackend`.

Now this is not necessary anymore and could be changed as you suggested.

Reply:
Done in PR #738